generate.py

```diff
@@ -29,53 +30,75 @@ def get_cmake_os(platform):
-            else:
-                raise RuntimeError('Unknown category: ' + category)
-            flags += xcode.GetCflags(configuration_name)
-
-            return flags
-        return get_flags
-    return get_factory
-
-def generic_flags_factories():
-    def get_factory(category):
-        def get_flags(configuration_name, configuration):
-            if configuration_name is None:
-                return []
-
-            flags = []
-            if category == 'c':
-                flags += configuration.get('cflags_c', [])
-            elif category == 'cc':
-                flags += configuration.get('cflags_cc', [])
-            else:
-                raise RuntimeError('Unknown category: ' + category)
-            flags += configuration.get('cflags', [])
+class Properties(object):
+    def __init__(self, category):
+        self._category = category
 
-            return flags
-        return get_flags
-    return get_factory
+    def compile_flags(self, configuration_name, configuration):
+        if configuration_name is None:
+            return []
 
-def get_flags_factories(platform, target):
-    if platform == 'Darwin':
-        return xcode_flags_factories(gyp.xcode_emulation.XcodeSettings(target))
+        flags = []
+        if self._category == 'c':
+            flags += configuration.get('cflags_c', [])
+        elif self._category == 'cc':
+            flags += configuration.get('cflags_cc', [])
+        else:
+            raise RuntimeError('Unknown category: ' + self._category)
+        flags += configuration.get('cflags', [])
+
+        return flags
+
+    @classmethod
+    def defines(self, configuration_name, configuration):
+        return configuration.get('defines', [])
+
+    @classmethod
+    def include_dirs(self, configuration_name, configuration):
+        return configuration.get('include_dirs', [])
+
+class EmulatedProperties(Properties):
+    def __init__(self, settings, category):
+        self._settings = settings
+        self._category = category
+
+    def compile_flags(self, configuration_name, configuration):
+        # TBD: should we call super(EmulatedProperties, self).compile_flags?
 
-    if platform == 'Windows':
-        # TBD: implement for win32
-        raise RuntimeError('Currently unsupported platform: ' + platform)
+        if configuration_name is None:
+            return []
 
-    return generic_flags_factories()
+        flags = []
+        if self._category == 'c':
+            flags += self._settings.GetCflagsC(configuration_name)
+        elif self._category == 'cc':
+            flags += self._settings.GetCflagsCC(configuration_name)
+        else:
+            raise RuntimeError('Unknown category: ' + self._category)
+        flags += self._settings.GetCflags(configuration_name)
+
+        return flags
+
+    def defines(self, configuration_name, configuration):
+        defines = Properties.defines(configuration_name, configuration)
```

Comment:
I believe `super` is preferable in case the hierarchy changes?

Reply:
I actually not sure which is more Pythonic. But seems like we should be using `@staticmethod` instead?

Either way, does `super` make sense. The way I see this code is that it is using some utility which _happens_ to be in `super` class namespace.

Reply:
I see. We can take that approach, but then I think we should remove the parent/child class relationship and forward all the methods explicitly.